## Project 3 Data Cleaning

In [2]:
# Dependencies
import pandas as pd
from pathlib import Path

In [3]:
# Use Pandas to the read data.
data_df = pd.read_csv("2023_data.csv")
data_df.head()

year  month carrier       carrier_name airport  \
0  2023     12      9E  Endeavor Air Inc.     ABE   
1  2023     12      9E  Endeavor Air Inc.     AEX   
2  2023     12      9E  Endeavor Air Inc.     AGS   
3  2023     12      9E  Endeavor Air Inc.     ALB   
4  2023     12      9E  Endeavor Air Inc.     ATL   

                                        airport_name  arr_flights  arr_del15  \
0  Allentown/Bethlehem/Easton, PA: Lehigh Valley ...         72.0        5.0   
1           Alexandria, LA: Alexandria International         62.0        7.0   
2        Augusta, GA: Augusta Regional at Bush Field         95.0       10.0   
3                   Albany, NY: Albany International         23.0        2.0   
4  Atlanta, GA: Hartsfield-Jackson Atlanta Intern...       2111.0      256.0   

   carrier_ct  weather_ct  ...  security_ct  late_aircraft_ct  arr_cancelled  \
0        2.46        1.00  ...          0.0              0.81            0.0   
1        4.25        0.00  ...          0.0              1.75            0.0   
2        5.94        0.00  ...          0.0              3.00            0.0   
3        0.56        0.00  ...          0.0              1.44            1.0   
4       76.88        8.75  ...          0.0            117.94            1.0   

   arr_diverted  arr_delay  carrier_delay  weather_delay  nas_delay  \
0           0.0      672.0           61.0          574.0       20.0   
1           0.0      348.0          252.0            0.0       33.0   
2           0.0      859.0          536.0            0.0       47.0   
3           0.0       75.0            9.0            0.0        0.0   
4           0.0    21424.0         8906.0          732.0     1487.0   

   security_delay  late_aircraft_delay  
0             0.0                 17.0  
1             0.0                 63.0  
2             0.0                276.0  
3             0.0                 66.0  
4             0.0              10299.0  

[5 rows x 21 columns]

In [4]:
data_df.count()

year                   22621
month                  22621
carrier                22621
carrier_name           22621
airport                22621
airport_name           22621
arr_flights            22574
arr_del15              22569
carrier_ct             22574
weather_ct             22574
nas_ct                 22574
security_ct            22574
late_aircraft_ct       22574
arr_cancelled          22574
arr_diverted           22574
arr_delay              22574
carrier_delay          22574
weather_delay          22574
nas_delay              22574
security_delay         22574
late_aircraft_delay    22574
dtype: int64

In [5]:
# Review all rows with null values
null_mask = data_df.isnull().any(axis=1)
null_rows = data_df[null_mask]

null_rows

year  month carrier                             carrier_name airport  \
779    2023     12      UA                 United Air Lines Network     BHM   
1254   2023     12      G7  GoJet Airlines LLC d/b/a United Express     SBN   
1423   2023     12      YV                       Mesa Airlines Inc.     ROA   
1470   2023     12      YX                         Republic Airline     ICT   
2278   2023     11      C5            CommuteAir LLC dba CommuteAir     BTV   
2311   2023     11      C5            CommuteAir LLC dba CommuteAir     PVD   
5472   2023     10      YV                       Mesa Airlines Inc.     PNS   
7901   2023      8      C5            CommuteAir LLC dba CommuteAir     DAY   
7914   2023      8      C5            CommuteAir LLC dba CommuteAir     LBB   
7929   2023      8      C5            CommuteAir LLC dba CommuteAir     PWM   
8284   2023      8      G7  GoJet Airlines LLC d/b/a United Express     CHS   
9209   2023      8      YV                       Mesa Airlines Inc.     ECP   
10184  2023      7      G7  GoJet Airlines LLC d/b/a United Express     MYR   
10192  2023      7      G7  GoJet Airlines LLC d/b/a United Express     RDU   
11096  2023      7      YV                       Mesa Airlines Inc.     LIT   
11644  2023      6      C5            CommuteAir LLC dba CommuteAir     CHS   
12047  2023      6      G7  GoJet Airlines LLC d/b/a United Express     CLT   
12129  2023      6      MQ                                Envoy Air     CLL   
12143  2023      6      MQ                                Envoy Air     EUG   
12147  2023      6      MQ                                Envoy Air     FSD   
12153  2023      6      MQ                                Envoy Air     GRB   
12159  2023      6      MQ                                Envoy Air     HPN   
12180  2023      6      MQ                                Envoy Air     MLI   
12202  2023      6      MQ                                Envoy Air     RIC   
12204  2023      6      MQ                                Envoy Air     ROC   
12205  2023      6      MQ                                Envoy Air     RST   
12534  2023      6      OO                    SkyWest Airlines Inc.     MLU   
12599  2023      6      OO                    SkyWest Airlines Inc.     SPI   
12671  2023      6      PT                        Piedmont Airlines     PWM   
12714  2023      6      QX                              Horizon Air     PSC   
12964  2023      6      YV                       Mesa Airlines Inc.     COS   
13722  2023      5      F9                        Frontier Airlines     BFF   
14992  2023      5      ZW              Air Wisconsin Airlines Corp     CVG   
15452  2023      4      C5            CommuteAir LLC dba CommuteAir     SYR   
15807  2023      4      G7  GoJet Airlines LLC d/b/a United Express     AVP   
16807  2023      4      YX                         Republic Airline     FSD   
16897  2023      4      ZW              Air Wisconsin Airlines Corp     PVD   
17755  2023      3      MQ                                Envoy Air     AZO   
17781  2023      3      MQ                                Envoy Air     DAY   
17795  2023      3      MQ                                Envoy Air     FNT   
18143  2023      3      OO                    SkyWest Airlines Inc.     LAW   
18270  2023      3      PT                        Piedmont Airlines     CWA   
18271  2023      3      PT                        Piedmont Airlines     DAY   
18591  2023      3      YV                       Mesa Airlines Inc.     BUR   
18631  2023      3      YV                       Mesa Airlines Inc.     MLI   
20514  2023      2      YV                       Mesa Airlines Inc.     FSM   
20555  2023      2      YV                       Mesa Airlines Inc.     PSP   
21282  2023      1      DL                  Delta Air Lines Network     TRI   
21496  2023      1      G7  GoJet Airlines LLC d/b/a United Express     BHM   
21511  2023      1      G7  GoJet Airlines LLC d/b/a Unit

#### 47 rows have all null (NaN) values for the fields containing flight data (columns arr_flights through late_aircraft_delay). These rows have no flight data at all so deleting them would not have a significant effect on the whole data set. 

####  5 rows have a null value in arr_del15 column only; in all 5 instances this is because there were only cancelled or diverted flights (no flights arrived, therefore the correct value is NA as there were no on-time arrivals or late arrivals). Because 5 out of 22621 data points is negligible, recommend also removing these 5 rows so the data set contains no null values.

In [7]:
# drop rows with null values
data_df = data_df.dropna()

In [8]:
# check columns counts to ensure they all match
data_df.count()

year                   22569
month                  22569
carrier                22569
carrier_name           22569
airport                22569
airport_name           22569
arr_flights            22569
arr_del15              22569
carrier_ct             22569
weather_ct             22569
nas_ct                 22569
security_ct            22569
late_aircraft_ct       22569
arr_cancelled          22569
arr_diverted           22569
arr_delay              22569
carrier_delay          22569
weather_delay          22569
nas_delay              22569
security_delay         22569
late_aircraft_delay    22569
dtype: int64

In [9]:
# drop unnecessary columns
data_df.columns

Index(['year', 'month', 'carrier', 'carrier_name', 'airport', 'airport_name',
       'arr_flights', 'arr_del15', 'carrier_ct', 'weather_ct', 'nas_ct',
       'security_ct', 'late_aircraft_ct', 'arr_cancelled', 'arr_diverted',
       'arr_delay', 'carrier_delay', 'weather_delay', 'nas_delay',
       'security_delay', 'late_aircraft_delay'],
      dtype='object')

In [10]:
data_df = data_df.drop(columns=['year', 'carrier'])
data_df.columns

Index(['month', 'carrier_name', 'airport', 'airport_name', 'arr_flights',
       'arr_del15', 'carrier_ct', 'weather_ct', 'nas_ct', 'security_ct',
       'late_aircraft_ct', 'arr_cancelled', 'arr_diverted', 'arr_delay',
       'carrier_delay', 'weather_delay', 'nas_delay', 'security_delay',
       'late_aircraft_delay'],
      dtype='object')

In [11]:
# rename columns to be more intuitive
data_df = data_df.rename(columns={
    'arr_flights':'total_arrivals',
    'arr_del15':'total_delays_ct',
    'nas_ct':'nat_air_sys_ct',
    'arr_cancelled':'flight_cancelled', 
    'arr_diverted':'flight_diverted',
    'arr_delay':'total_delays_min',
    'carrier_delay':'carrier_delay_min',
    'weather_delay': 'weather_delay_min',
    'nas_delay':'nat_air_sys_delay_min',
    'security_delay':'security_delay_min',
    'late_aircraft_delay':'late_aircraft_delay_min'   

    })
data_df.head()

month       carrier_name airport  \
0     12  Endeavor Air Inc.     ABE   
1     12  Endeavor Air Inc.     AEX   
2     12  Endeavor Air Inc.     AGS   
3     12  Endeavor Air Inc.     ALB   
4     12  Endeavor Air Inc.     ATL   

                                        airport_name  total_arrivals  \
0  Allentown/Bethlehem/Easton, PA: Lehigh Valley ...            72.0   
1           Alexandria, LA: Alexandria International            62.0   
2        Augusta, GA: Augusta Regional at Bush Field            95.0   
3                   Albany, NY: Albany International            23.0   
4  Atlanta, GA: Hartsfield-Jackson Atlanta Intern...          2111.0   

   total_delays_ct  carrier_ct  weather_ct  nat_air_sys_ct  security_ct  \
0              5.0        2.46        1.00            0.73          0.0   
1              7.0        4.25        0.00            1.00          0.0   
2             10.0        5.94        0.00            1.06          0.0   
3              2.0        0.56        0.00            0.00          0.0   
4            256.0       76.88        8.75           52.43          0.0   

   late_aircraft_ct  flight_cancelled  flight_diverted  total_delays_min  \
0              0.81               0.0              0.0             672.0   
1              1.75               0.0              0.0             348.0   
2              3.00               0.0              0.0             859.0   
3              1.44               1.0              0.0              75.0   
4            117.94               1.0              0.0           21424.0   

   carrier_delay_min  weather_delay_min  nat_air_sys_delay_min  \
0               61.0              574.0                   20.0   
1              252.0                0.0                   33.0   
2              536.0                0.0                   47.0   
3                9.0                0.0                    0.0   
4             8906.0              732.0                 1487.0   

   security_delay_min  late_aircraft_delay_min  
0                 0.0                     17.0  
1                 0.0                     63.0  
2                 0.0                    276.0  
3                 0.0                     66.0  
4                 0.0                  10299.0

In [17]:
# Move city column to position following airport column
data_df = data_df[['month', 'carrier_name', 'airport', 'city', 'total_arrivals', 'total_delays_ct',
                   'carrier_ct', 'weather_ct', 'nat_air_sys_ct', 'security_ct', 'late_aircraft_ct',
                   'flight_cancelled', 'flight_diverted', 'total_delays_min', 'carrier_delay_min',
                   'weather_delay_min', 'nat_air_sys_delay_min', 'security_delay_min', 'late_aircraft_delay_min']]

# Print the first few rows of the 'city' column for debugging
print("First 20 rows of the 'city' column:")
print(data_df['city'].head(20))

# Define a function to split city and state
def split_city_state(city_state_str):
    parts = city_state_str.rsplit(',', 1)  # Split on the last comma
    if len(parts) == 2:
        city = parts[0].replace('/', ', ').strip()
        state = parts[1].strip()
        print(f"Processing: {city_state_str} -> City: '{city}', State: '{state}'")  # Debug print
        return city, state
    print(f"Could not split: {city_state_str}")  # Debug print for unsplit cases
    return city_state_str, ''  # Handle cases with no comma

# Apply the function to split city and state
data_df['city'], data_df['state'] = zip(*data_df['city'].apply(split_city_state))

# Verify the columns and content
print("After splitting 'city' and 'state':")
print(data_df[['city', 'state']].head(20))

# Rearrange columns
data_df = data_df[['month', 'carrier_name', 'airport', 'city', 'state', 'total_arrivals', 'total_delays_ct',
                   'carrier_ct', 'weather_ct', 'nat_air_sys_ct', 'security_ct', 'late_aircraft_ct', 
                   'flight_cancelled', 'flight_diverted', 'total_delays_min', 'carrier_delay_min', 
                   'weather_delay_min', 'nat_air_sys_delay_min', 'security_delay_min', 'late_aircraft_delay_min']]

# Write the cleaned data to a new CSV file
data_df.to_csv("2023_data_cleaned_sh.csv", index=False, header=True)

# Display the final DataFrame
print("\nFinal DataFrame:")
print(data_df.head())

First 20 rows of the 'city' column:
0     Allentown/Bethlehem/Easton, PA
1                     Alexandria, LA
2                        Augusta, GA
3                         Albany, NY
4                        Atlanta, GA
5                       Appleton, WI
6                         Austin, TX
7                      Asheville, NC
8                      Kalamazoo, MI
9                       Hartford, CT
10                    Binghamton, NY
11                        Bangor, ME
12                    Birmingham, AL
13            Bloomington/Normal, IL
14                     Nashville, TN
15                        Boston, MA
16                   Baton Rouge, LA
17                    Burlington, VT
18                       Buffalo, NY
19                     Baltimore, MD
Name: city, dtype: object
Processing: Allentown/Bethlehem/Easton, PA -> City: 'Allentown, Bethlehem, Easton', State: 'PA'
Processing: Alexandria, LA -> City: 'Alexandria', State: 'LA'
Processing: Augusta, GA -> City: 'August

In [12]:
#split data in airport_name into two columns
data_df[['city', 'name']] = data_df['airport_name'].str.split(': ', expand=True)
data_df.head()

month       carrier_name airport  \
0     12  Endeavor Air Inc.     ABE   
1     12  Endeavor Air Inc.     AEX   
2     12  Endeavor Air Inc.     AGS   
3     12  Endeavor Air Inc.     ALB   
4     12  Endeavor Air Inc.     ATL   

                                        airport_name  total_arrivals  \
0  Allentown/Bethlehem/Easton, PA: Lehigh Valley ...            72.0   
1           Alexandria, LA: Alexandria International            62.0   
2        Augusta, GA: Augusta Regional at Bush Field            95.0   
3                   Albany, NY: Albany International            23.0   
4  Atlanta, GA: Hartsfield-Jackson Atlanta Intern...          2111.0   

   total_delays_ct  carrier_ct  weather_ct  nat_air_sys_ct  security_ct  ...  \
0              5.0        2.46        1.00            0.73          0.0  ...   
1              7.0        4.25        0.00            1.00          0.0  ...   
2             10.0        5.94        0.00            1.06          0.0  ...   
3              2.0        0.56        0.00            0.00          0.0  ...   
4            256.0       76.88        8.75           52.43          0.0  ...   

   flight_cancelled  flight_diverted  total_delays_min  carrier_delay_min  \
0               0.0              0.0             672.0               61.0   
1               0.0              0.0             348.0              252.0   
2               0.0              0.0             859.0              536.0   
3               1.0              0.0              75.0                9.0   
4               1.0              0.0           21424.0             8906.0   

   weather_delay_min  nat_air_sys_delay_min  security_delay_min  \
0              574.0                   20.0                 0.0   
1                0.0                   33.0                 0.0   
2                0.0                   47.0                 0.0   
3                0.0                    0.0                 0.0   
4              732.0                 1487.0                 0.0   

   late_aircraft_delay_min                            city  \
0                     17.0  Allentown/Bethlehem/Easton, PA   
1                     63.0                  Alexandria, LA   
2                    276.0                     Augusta, GA   
3                     66.0                      Albany, NY   
4                  10299.0                     Atlanta, GA   

                                       name  
0               Lehigh Valley International  
1                  Alexandria International  
2            Augusta Regional at Bush Field  
3                      Albany International  
4  Hartsfield-Jackson Atlanta International  

[5 rows x 21 columns]

In [13]:
#drop airport_name and name columns 
data_df = data_df.drop(columns=['airport_name', 'name'])
data_df.columns

Index(['month', 'carrier_name', 'airport', 'total_arrivals', 'total_delays_ct',
       'carrier_ct', 'weather_ct', 'nat_air_sys_ct', 'security_ct',
       'late_aircraft_ct', 'flight_cancelled', 'flight_diverted',
       'total_delays_min', 'carrier_delay_min', 'weather_delay_min',
       'nat_air_sys_delay_min', 'security_delay_min',
       'late_aircraft_delay_min', 'city'],
      dtype='object')

In [14]:
# move city column to position following airport column
data_df = data_df[['month', 'carrier_name', 'airport', 'city', 'total_arrivals', 'total_delays_ct',
       'carrier_ct', 'weather_ct', 'nat_air_sys_ct', 'security_ct',
       'late_aircraft_ct', 'flight_cancelled', 'flight_diverted', 'total_delays_min',
       'carrier_delay_min', 'weather_delay_min', 'nat_air_sys_delay_min',
       'security_delay_min', 'late_aircraft_delay_min']]
data_df.head()

month       carrier_name airport                            city  \
0     12  Endeavor Air Inc.     ABE  Allentown/Bethlehem/Easton, PA   
1     12  Endeavor Air Inc.     AEX                  Alexandria, LA   
2     12  Endeavor Air Inc.     AGS                     Augusta, GA   
3     12  Endeavor Air Inc.     ALB                      Albany, NY   
4     12  Endeavor Air Inc.     ATL                     Atlanta, GA   

   total_arrivals  total_delays_ct  carrier_ct  weather_ct  nat_air_sys_ct  \
0            72.0              5.0        2.46        1.00            0.73   
1            62.0              7.0        4.25        0.00            1.00   
2            95.0             10.0        5.94        0.00            1.06   
3            23.0              2.0        0.56        0.00            0.00   
4          2111.0            256.0       76.88        8.75           52.43   

   security_ct  late_aircraft_ct  flight_cancelled  flight_diverted  \
0          0.0              0.81               0.0              0.0   
1          0.0              1.75               0.0              0.0   
2          0.0              3.00               0.0              0.0   
3          0.0              1.44               1.0              0.0   
4          0.0            117.94               1.0              0.0   

   total_delays_min  carrier_delay_min  weather_delay_min  \
0             672.0               61.0              574.0   
1             348.0              252.0                0.0   
2             859.0              536.0                0.0   
3              75.0                9.0                0.0   
4           21424.0             8906.0              732.0   

   nat_air_sys_delay_min  security_delay_min  late_aircraft_delay_min  
0                   20.0                 0.0                     17.0  
1                   33.0                 0.0                     63.0  
2                   47.0                 0.0                    276.0  
3                    0.0                 0.0                     66.0  
4                 1487.0                 0.0                  10299.0

In [15]:
# write cleaned data to new csv file
data_df.to_csv("2023_data_cleaned.csv", index=False, header=True)

PermissionError: [Errno 13] Permission denied: '2023_data_cleaned.csv'

#### 2023_data_cleaned.csv will be imported into pgAdmin SQL database for further data exploration